src/clj/puppetlabs/services/jruby_pool_manager/impl/instance_pool.clj

```diff
@@ -0,0 +1,65 @@
+(ns puppetlabs.services.jruby-pool-manager.impl.instance-pool
+  (:require [puppetlabs.services.jruby-pool-manager.impl.jruby-agents :as jruby-agents]
+            [puppetlabs.services.protocols.jruby-pool :as pool-protocol]
+            [puppetlabs.services.jruby-pool-manager.impl.jruby-internal :as jruby-internal]
+            [puppetlabs.services.jruby-pool-manager.jruby-schemas :as jruby-schemas]
+            [clojure.tools.logging :as log]
+            [puppetlabs.i18n.core :as i18n])
+  (:import (puppetlabs.services.jruby_pool_manager.jruby_schemas InstancePool)))
+
+(extend-type InstancePool
+  pool-protocol/JRubyPool
+  (fill
+    [this]
+    (let [modify-instance-agent (jruby-agents/get-modify-instance-agent this)]
+      (jruby-agents/send-agent modify-instance-agent
+                               #(jruby-agents/prime-pool! this))))
+
+  (shutdown
+    [this]
+    (jruby-agents/flush-pool-for-shutdown! this))
+
+  (lock
+    [this]
+    (let [pool (jruby-internal/get-pool this)]
+      (.lock pool)))
+
+  (lock-with-timeout
+    [this timeout time-unit]
+    (let [pool (jruby-internal/get-pool this)]
+      (.lockWithTimeout pool timeout time-unit)))
+
+  (unlock
+    [this]
+    (let [pool (jruby-internal/get-pool this)]
+      (.unlock pool)))
+
+  (borrow
+    [this]
+    (jruby-internal/borrow-from-pool this))
+
+  (borrow-with-timeout
+    [this timeout]
+    (jruby-internal/borrow-from-pool-with-timeout this timeout))
+
+  (return
+    [this instance]
+    (when (jruby-schemas/jruby-instance? instance)
```

Comment:
When would this not be a jruby instance?

Reply:
It might be a pill, if this thread tried to borrow when a pill had been inserted. I actually added this when I found a test that verified that when a pill is inserted, no error is thrown. The Java `return` method will error if the thing being returned is not the registered instance, see https://github.com/puppetlabs/jruby-utils/pull/78/files#diff-4486a0d17a50859d82ce51b87b75c9d5R256. In practice, the pill being returned is most likely also the one that the pool is holding onto, but in this test, it was a new one that hadn't actually been inserted into the pool yet.

Reply:
so to make sure I'm understanding - the only reason you don't also call `return` on pills is so you can avoid that error in a test where the pill was never registered?

Reply:
This mirrors the original implementation, and it looks like the [simple commit](https://github.com/puppetlabs/jruby-utils/commit/4bf607037e00e74b3be2185e10b04e842e3c5ba6) that added that is part of a very fraught PR: https://github.com/puppetlabs/jruby-utils/pull/34. Such a case might never arise in the wild (that something that is not an instance or the registered pill might be returned), but I wasn't sure, so I just added the same logic. If we think it's safe to remove here (and let the logic in the Java handle it), then we should also remove it (and the weird test) from the original implementation.

Reply:
I'm generally a big fan of removing things, but don't necessarily have strong feeling right here right now. Mostly wanted to make sure I was understanding things.

Reply:
I didn't totally follow this thread, but what I got from it is that this protocol method `return` can take a jruby-instance or poison-pill, but none of this logic is applicable to poison-pills, so if poison-pill (eg not a jruby-instance) be a no-op. Is that accurate?